### PR TITLE
Product-Deploy Week 33

### PR DESCRIPTION
**Pull request by CZYK**
increment for: https://github.com/studyportals/product-deploy/issues/30

### CODE REVIEW

File: lib/configure.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * @private
+ */
+const _copyConfig = (opts) =>{
+
+	return new Promise((resolve) =>{
+
+		const defaults = {
+			'product': path.basename(process.cwd()),
+			'to': null,
+			'env': env.PROD,
```
Comment (iliqportals) [inline lib/configure.js]: would be safer to default to TST

File: lib/configure.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		const productType = _products[opts.product];
+
+		let configGlobs = [];
+
+		if(productType === 'CMS' || productType === 'Academic'){
+
+			configGlobs.push(path.join(environmentConfigRoot, 'Base', productType) + '/**');
+		}
+
+		configGlobs.push(path.join(environmentConfigRoot, opts.product) + '/**');
```
Comment (iliqportals) [inline lib/configure.js]: path.join(environmentConfigRoot, opts.product) + '/**'
path.join(environmentConfigRoot, opts.product, "**")
Act the same, but avoids using different operators

File: lib/env.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+"use strict";
+
+/**
+ * Environments
+ */
+const env = {
+	'DEV': 'Development',
+	'TST': 'Testing',
+	'STG': 'Staging',
+	'PROD': 'Production',
```
Comment (CZYK) [inline lib/env.js]: Awesome, will do!

File: lib/env.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+	'LIVE': 'Live',
```
Comment (CZYK) [inline lib/env.js]: I already though this was here for legacy reasons.

File: lib/configure.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	PhD: 'Academic',
+	ShortCourses: 'Academic',
+	UniversityAdmin: 'Academic',
+	Scholarship: 'CMS',
+	STeXX: 'CMS',
+	DistanceLearning: 'Academic',
+	Statistics: 'Academic',
+	ServiceLayer: 'GenY',
+	Publish: 'CMS',
+	LanguageLearning: 'CMS',
```
Comment (CZYK) [inline lib/configure.js]: I totally agree! I want to go to a situation where there is not even a difference  between Academic, CMS, GenY and maybe even skip the Base config. I can remove this but that also requires moving around the Local config folder in our products. That might be a step to far for now or do you think it is good to do so as well?

File: lib/attachToGulp.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * @param {string} opts.buildDir
+ * @param {string} opts.env
+ * @param {string} opts.enableCompression
+ */
+const attachToGulp = (gulp, opts) =>{
+
+	const PD = require('./../index');
+	gulp.task('spici.prepare', () => PD.prepare.ensureEmptyDir(opts.buildDir));
+	gulp.task('spici.composer', () => PD.composer.install());
+	gulp.task('spici.bower', () => PD.bower.install());
```
Comment (CZYK) [inline lib/attachToGulp.js]: Fair enough! Will disable/deprecate it in the same way as SiteDB and remove it in the next major release.

File: lib/attachToGulp.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	}));
+	gulp.task('spici.uglify', () => PD.js.compile({
+		from: [
+			`${opts.buildDir}/**/*.js`,
+			`!${opts.buildDir}/node_modules/**/*`,
+		],
+		to: opts.buildDir,
+		uglify: opts.enableCompression
+	}));
+	gulp.task('spici.siteDB', done => { done() });
```
Comment (CZYK) [inline lib/attachToGulp.js]: Sounds cool :)

File: lib/attachToGulp.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	gulp.task('spici.configure', function(){
+		return PD.configure({
+			to: opts.buildDir,
+			env: opts.env
+		});
+	});
+	// // TODO: spici.assemble
+	gulp.task('spici.replace', done => { done() });
```
Comment (CZYK) [inline lib/attachToGulp.js]: Note that we still use `prtl-registry` to invoke our main deploys. Those main deploys invoke these tasks. That's why I overwrite them for now with an empty task. After all the steps are moved to this repo we can remove these tasks.

File: lib/configure.js (hunk 5 above)
Comment (iliqportals) [inline lib/configure.js]: i think we can do it, i'll ge to it now
Talked with Rob and he confirmed that the Academic, GenY, Base have no longer any value.

The configuration step will be now : Copy * from Deploy\Config\[ENVIRONMENT]\** into buildDir
I'll do the change in the current spici package

File: lib/attachToGulp.js (hunk 8 above)
Comment (iliqportals) [inline lib/attachToGulp.js]: I have removed the depricated tasks in spici 4.0
spici 4.0 will be applied to all portals tomorrow

File: lib/configure.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+			'to': null,
+			'env': env.LIVE,
+		};
+		opts = Object.assign(defaults, opts);
+
+		const promises = [];
+
+		promises.push(_copyConfig({to: opts.to}));
+
+		if(opts.env !== env.LIVE || env.PROD){
```
Comment (iliqportals) [inline lib/configure.js]: [!] This is always true
env.PROD = Production (constantly)
env.PROD ? true : false => true

if(opts.env !== env.LIVE || env.PROD){
I think you meant to write 
if(opts.env !== env.LIVE && opts.env !== env.PROD) (any Environment except Production)

File: lib/configure.js (hunk 5 above)
Comment (iliqportals) [inline lib/configure.js]: *
in the Case of SL
opts.env = Live/Private | Live/Public

File: lib/attachToGulp.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+			opts.buildDir + '/**/*.scss',
+			'!' + opts.buildDir + '/bower_components/**/*',
+			'!' + opts.buildDir + '/node_modules/**/*',
+		],
+		to: opts.buildDir,
+		outputStyle: opts.enableCompression ? PD.sass.OUTPUTSTYLE.COMPRESSED : PD.sass.OUTPUTSTYLE.NESTED,
+	}));
+	gulp.task('spici.uglify', () => PD.js.compile({
+		from: [
+			`${opts.buildDir}/**/*.js`,
```
Comment (CZYK) [inline lib/attachToGulp.js]: Do you remember that you files an issue where you explicitly requested to have the bower_components also uglified?

https://github.com/studyportals/product-deploy/issues/23

File: lib/attachToGulp.js (hunk 8 above)
Comment (CZYK) [inline lib/attachToGulp.js]: Sounds good, in that case I can remove them completely next week.

File: lib/attachToGulp.js (hunk 10 above)
Comment (CZYK) [inline lib/attachToGulp.js]: `bower_components` is now also excluded as discussed with you on Slack.

File: lib/configure.js (hunk 5 above)
Comment (CZYK) [inline lib/configure.js]: Awesome, will adapt it as well in my code. For now I only take the portal deploys into account, once we start using these packages for ServiceLayer we can change it accordingly.

File: lib/configure.js (hunk 1 above)
Comment (CZYK) [inline lib/configure.js]: I have made `Testing` the default env in the exported `configure` method.

File: lib/configure.js (hunk 2 above)
Comment (CZYK) [inline lib/configure.js]: Even better:
`path.join(environmentConfigRoot, opts.product, "**")`

**Pull request merged**